File: src/layout/widget/controller/io/InOutput.java
package layout.widget.controller.io;

import java.util.UUID;

import layout.area.IOLayer;

/**
 * Special {XPut}, a composition of {@link layout.widget.controller.io.Input} and
 * {@link layout.widget.controller.io.Output}.
 */
public class InOutput<T> implements XPut<T> {
	public final Input<T> i;
	public final Output<T> o;

	public InOutput(UUID id, String name, Class<? super T> c) {
		o = new Output<>(id, name, c);
		i = new Input<>(name, c, o::setValue) {
			// This solves the problem of binding ot itself and this can be
			// used to notify the graphics as well, so its on a user level - he
			// can be prevented from even trying.
			@Override
			public boolean canBind(Output<?> output) {
				return output!=o && super.canBind(output);
			}
		};

		IOLayer.all_inoutputs.add(this);
	}
}

File: src/player/main/sp/it/pl/layout/widget/controller/io/InOutput.kt
package sp.it.pl.layout.widget.controller.io

import java.util.UUID
import sp.it.util.type.VType
import sp.it.util.type.type

/** [XPut], that is a composition of [Input] and [Output]. */
class InOutput<T>: XPut<T> {
   @JvmField val o: Output<T>
   @JvmField val i: Input<T>

   constructor(id: UUID, name: String, type: VType<T>, initialValue: T) {
      this.o = Output(id, name, type, initialValue)
      this.i = object: Input<T>(name, type, initialValue, { o.value = it }) {
         override fun isAssignable(output: Output<*>) = output!==o && super.isAssignable(output)
      }
   }

   fun appWide() = apply {
      IOLayer.allInoutputs += this
   }

   companion object {
      inline operator fun <reified T> invoke(id: UUID, name: String, initialValue: T): InOutput<T> = InOutput(id, name, type(), initialValue)
   }

}